### DISCUSSION TITLE
请问如何加载pkg包的txt文件

### DISCUSSION THREAD

**Discussion by constly** [ejoy/ant#38] [Q&A] [answered]
比如在ant.resources下面存在一个文件叫 abc.txt，那么我在其他地方应该如何加载这个文件呢？
我尝试使用:

` io.open("/pkg/ant.resources/abc.txt")`

然后程序报错说：

`/pkg/ant.resources/abc.txt: No such file or directory`

我也尝试用assetmgr，最终也没整明白 = =

**Comment by actboy168** [comment]
对于运行时的文件的扩展名有白名单机制。txt不是白名单上的扩展名，换个在白名单上的扩展名，例如abc.ant，就行了。

**Reply by constly** [reply]
感谢回答，另外，这个“支持的扩展名”有地方配置吗，如果不能有自己的数据格式，实在是不方便

**Reply by actboy168** [reply]
你可以让某个目录忽略这些规则。
在项目的根目录添加.vfsignore文件
``` yml
ignore:
    /pkg/ant.resources
```

**Reply by constly** [reply]
请问确定是用 io.open 打开吗，我把文件名改为“abc.ant”，也还是找不到，ignore配了好像也没啥效果，头秃 = =

**Reply by actboy168** [reply, marked as answer]
``` lua
local aio = import_package "ant.io"
print(aio.readall_s "/pkg/ant.font/main.lua")
```

**Reply by constly** [reply]
哇塞，非常感谢，终于读取成功啦！
对于.vfsignore，我开始理解为放引擎的根目录了，调试后才发现，
比如我的启动命令是：
```
bin/msvc/release/lua.exe game/entry.lua
```
.vfsignore文件应该放在game目录下。

经过一些尝试，我发现一个问题，比如我有个包叫: game.com.res，
里面分别有三个文件：abc.lua, abc.txt, miner-1.glb(这个文件为引擎自带资源)，又配置了
```
ignore:
    /pkg/game.com.res
```
会出现miner-1.glb加载失败的情况，如果不配置.vfsignore是可以正常加载的，其中报错如下：
```
[2024-01-20 11:00:19.78][INFO ]( ant.window|world )(F:/ant/game/pkg/game.editor.entry/init_system.lua:38) this is abc.lua
[2024-01-20 11:00:19.78][INFO ]( ant.window|world )(F:/ant/game/pkg/game.editor.entry/init_system.lua:41) this is abc.txt
[2024-01-20 11:00:19.78][ERROR]( ant.window|world )`read `/pkg/game.com.res/resource/miner-1.glb|mesh.prefab` failed.
stack traceback:
        ( service:9 )
        F:/ant/pkg/ant.io/main.lua:7: in upvalue 'readall'
        F:/ant/pkg/ant.io/main.lua:12: in field 'readall'
        F:/ant/pkg/ant.ecs/main.lua:187: in upvalue 'create_template'
        F:/ant/pkg/ant.ecs/main.lua:225: in upvalue 'prefab_instance'
        F:/ant/pkg/ant.ecs/main.lua:257: in method '_flush_instance_queue'
        F:/ant/pkg/ant.ecs/entity_system.lua:71: in local 'f'
        F:/ant/pkg/ant.ecs/main.lua:389: in function <F:/ant/pkg/ant.ecs/main.lua:384>
        F:/ant/pkg/ant.ecs/main.lua:573: in upvalue 'system_changed'
        F:/ant/pkg/ant.ecs/main.lua:579: in method 'pipeline_init'
        F:/ant/pkg/ant.window/service/world.lua:80: in upvalue 'f'
```
加载代码如下:
```
function m:init_world()
    local prefab = world:create_instance {
        prefab = "/pkg/game.com.res/resource/miner-1.glb|mesh.prefab",
        on_ready = function ()
            local mq = w:first "main_queue camera_ref:in"
            local ce <close> = world:entity(mq.camera_ref, "camera:in")
            local dir = math3d.vector(0, -1, 1)
            if not icamera.focus_prefab(ce, entities, dir) then
                error "aabb not found"
            end
        end
    }
    entities = prefab.tag['*']
end
```
这怎么理解呢~

**Reply by actboy168** [reply]
ignore会将所有的vfs规则都忽略，而glb文件是依赖此变为一个可以在运行时加载的格式。所以你不能吧glb文件放在被忽略的目录里。

**Reply by cloudwu** [reply]
ignore 把一个目录变成了普通目录，里面所有文件都成了普通文件。

但是 .glb 是需要魔法才能用的，没有魔法它就是一个普通的二进制数据块。

https://github.com/ejoy/ant/wiki/Asset

**Reply by constly** [reply]
好的，了解，这块我想我需要多研究下文档和源码，感谢两位大佬的回复。

**Reply by wk34u** [reply]
楼主是想实现 `game.com.res` 包下有 `abc.lua, abc.txt, miner-1.glb` 并且即可以通过 `aio.readall_s` 读取 `*.txt`  又可以通过`world:create_instance` 使用 `*.glb`?

**Reply by constly** [reply]
是的，能这样更好，因为我想一个编辑器，编辑器既需要加载自己的文件格式，也需要能加载引擎的文件。
如果这些不能放在一个目录，会比较麻烦。

**Reply by wk34u** [reply]
https://github.com/ejoy/ant/blob/master/pkg/ant.vfs/tiny.lua#L23-L56 看起来目前引擎能接受的扩展名都在这里, 在不修改引擎代码的前提下，楼主或许可以用 `abc.txt.ant` 的方法来曲线救国?

**Reply by constly** [reply]
噢，我也看到这些了，目前确实不想动到引擎。我最近先看看其他会遇到的问题，文件后缀先放一边。
感谢你的回复，周末愉快！

**Reply by cloudwu** [reply]
编辑器是一种比较特殊的应用，因为它是在创造数据。它不太适合 vfs 现在的文件映射方式，vfs 提供的是只读的文件系统。

所以做编辑器最好是绕开 vfs 直接用原生系统 io 读取本地文件。如果后面你想让这个文件出现在 vfs 中，可以用 vfs 的内存接口。即：把外部文件在运行时导入 vfs 。

见 https://github.com/ejoy/ant/blob/master/test/vfsmem/pkg/test/service/main.lua

**Reply by cloudwu** [reply]
换句话说，你可以用 io.open 打开任何本地目录下的文件。然后可以把数据内容写到 vfs 的任意路径上。这个数据仅存在于当前进程的内存，不会对持久化数据有任何修改。

然后，所有引擎其它部分都可以通过 vfs api 访问到它。

**Reply by constly** [reply]
OK，OK，明白了~
